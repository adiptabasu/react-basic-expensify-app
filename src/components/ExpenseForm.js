import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
// import '../../node_modules/react-dates/lib/css/_datepicker.css';
const now = moment();

class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calenderFocused: false,
            error: ''
        }
        this.onDescriptionChange = this.onDescriptionChange.bind(this);
        this.onNoteChange = this.onNoteChange.bind(this);
        this.onAmountChange = this.onAmountChange.bind(this);
        this.onDateChange = this.onDateChange.bind(this);
        this.onCalendarFocusChange = this.onCalendarFocusChange.bind(this);
        this.onFormSubmit = this.onFormSubmit.bind(this);
    }
    onDescriptionChange(e) {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }
    onAmountChange = (e) => {
        const amount = e.target.value;
        const matchVal = amount.match(/^\d+\.{0,1}\d{0,2}$/);
        if (!!matchVal || !amount)
            this.setState(() => ({ amount }));
        else
            this.setState((prevState) => {
                amount: prevState.amount
            })
    }
    onDateChange = (createdAt) => {
        if (createdAt)
            this.setState(() => ({ createdAt }));
    }
    onCalendarFocusChange = ({ focused }) => {
        this.setState({ calenderFocused: focused });
    }
    onFormSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Description and Amount are Mandatory' }));
        }
        else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    }
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onFormSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="number"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calenderFocused}
                        onFocusChange={this.onCalendarFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={(day) => (false)}
                    />
                    <textarea
                        placeholder="Add a Note"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    ></textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        );
    }
}

export default ExpenseForm;