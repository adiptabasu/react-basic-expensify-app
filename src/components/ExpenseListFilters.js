import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

class ExpenseListFilters extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            calendarFocused: null
        }
        this.onDatesChange = this.onDatesChange.bind(this);
        this.onFocusChange = this.onFocusChange.bind(this);
    }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.dispatch(setStartDate({ startDate }));
        this.props.dispatch(setEndDate({ endDate }));
    }
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    }
    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input
                            className="text-input"
                            type="text"
                            value={this.props.filters.text} onChange={(e) => {
                                // console.log(e.target.value);
                                this.props.dispatch(setTextFilter({ text: e.target.value }));
                            }}
                            placeholder="Search Expenses"
                        />
                    </div>
                    <div className="input-group__item">
                        <select className="select" value={this.props.sortBy} onChange={(e) => {
                            let sortVal = e.target.value;
                            if (sortVal == 'date') {
                                this.props.dispatch(sortByDate());
                            }
                            else if (sortVal == 'amount') {
                                this.props.dispatch(sortByAmount());
                            }
                        }}>
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            numberOfMonths={1}
                            isOutsideRange={() => (false)}
                            showClearDates={true}
                        />
                    </div>
                </div>


            </div>
        );
    }
}

// const ExpenseListFilters = (props) => (
//     <div>
//         <input type="text" value={props.filters.text} onChange={(e) => {
//             // console.log(e.target.value);
//             props.dispatch(setTextFilter({ text: e.target.value }));
//         }} />
//         <select value={props.sortBy} onChange={(e) => {
//             let sortVal = e.target.value;
//             if (sortVal == 'date') {
//                 props.dispatch(sortByDate());
//             }
//             else if (sortVal == 'amount') {
//                 props.dispatch(sortByAmount());
//             }
//         }}>
//             <option value="date">Date</option>
//             <option value="amount">Amount</option>
//         </select>
//     </div>
// );

const mapStateToProps = (state) => {
    // console.log(state.filters);
    return { filters: state.filters }
}

const ConnectedExpenseListFilters = connect(mapStateToProps)(ExpenseListFilters);

export default ConnectedExpenseListFilters;