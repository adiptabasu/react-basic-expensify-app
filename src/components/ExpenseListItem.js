import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { removeExpense } from '../actions/expenses';

const ExpenseListItem = (props) => {
    let removeItem = () => {
        props.dispatch(removeExpense({ id: props.id }));
    };
    return (
        <div>
            <Link to={`/edit/${props.id}`}>
                <h3>
                    {props.description}
                </h3>
            </Link>
            <p>
                {props.amount}-{props.createdAt}
            </p>
            <button onClick={removeItem}>Remove</button>
        </div>
    );
};

const ConnectedExpenseListItem = connect()(ExpenseListItem);

export default ConnectedExpenseListItem;

export { ExpenseListItem };