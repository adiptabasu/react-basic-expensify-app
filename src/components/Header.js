import React from 'react';
import { NavLink } from 'react-router-dom'

const Header = () => (
    <div>
        <header>
            <h1>Expensify</h1>
            <NavLink
                to="/"
                // activeClassName="is-active"
                className={(props) => ("nav-link " + (props.isActive ? "is-active" : undefined))}
            >Home</NavLink>
            <NavLink
                to="/create"
                // activeClassName="is-active"
                className={(props) => ("nav-link " + (props.isActive ? "is-active" : undefined))}
            >Create Expense</NavLink>
            <NavLink
                to="/edit"
                // activeClassName="is-active"
                className={(props) => ("nav-link " + (props.isActive ? "is-active" : undefined))}
            >Edit</NavLink>
            <NavLink
                to="/help"
                // activeClassName="is-active"
                className={(props) => ("nav-link " + (props.isActive ? "is-active" : undefined))}
            >Help</NavLink>
        </header>
    </div>
);

export default Header;