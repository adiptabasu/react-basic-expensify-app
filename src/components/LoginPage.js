import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';
// import { useNavigate } from 'react-router-dom';
import { bindActionCreators } from 'redux';

const LoginPage = (props) => {
    // console.log(props);
    // const navigate = useNavigate();
    return (
        <div className="box-layout">
            <div className="box-layout__box">
                <h1 className="box-layout__title">Expensify</h1>
                <p>It's time to get your expenses under control</p>
                <button className="button" onClick={props.startLogin}>Login with Google</button>
            </div>
        </div>
    );
};

const mapDispatchToProps = (dispatch) => ({
    // startLogin,
    // dispatch
    startLogin: () => dispatch(startLogin())
});

const ConnectedLoginPage = connect(undefined, mapDispatchToProps)(LoginPage);

export { ConnectedLoginPage as default, LoginPage };