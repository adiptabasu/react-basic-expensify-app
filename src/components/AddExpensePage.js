import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';
import { useNavigate } from 'react-router-dom';

const AddExpensePage = (props) => {
    const navigate = useNavigate();
    return (
        <div>
            <h1>Add Expense</h1>
            <ExpenseForm
                onSubmit={(expense) => {
                    // console.log('Inside AddExpensePage');
                    // props.dispatch(addExpense(expense));
                    props.onSubmit(expense);
                    navigate('/');
                }}
            />
        </div>
    );
};

const mapDispatchToProps = (dispatch) => (({ onSubmit: (expense => dispatch(addExpense(expense))) }))

const ConnectedAddExpensePage = connect(undefined, mapDispatchToProps)(AddExpensePage);

export default ConnectedAddExpensePage;