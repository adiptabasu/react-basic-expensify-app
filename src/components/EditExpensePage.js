import React from 'react';
import { useNavigate, useParams } from 'react-router-dom';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRomoveExpense } from '../actions/expenses';

const EditExpensePage = (props) => {
    let params = useParams();
    const navigate = useNavigate();
    const expenseSelected = props.expenses.find((exp) => (exp.id == params.id));
    console.log(expenseSelected);
    return (<div>
        <ExpenseForm
            expense={expenseSelected}
            onSubmit={(expense) => {
                console.log('updated', expense);
                props.dispatch(startEditExpense({ id: params.id, updates: expense }));
                navigate('/');
            }}
        />
        <button onClick={() => {
            props.dispatch(startRomoveExpense({ id: params.id }));
            navigate('/');
        }}>Remove</button>
    </div>);
};

const mapStateToProps = (state) => {
    return ({
        expenses: state.expenses
    });
}

const ConnectedEditExpensePage = connect(mapStateToProps)(EditExpensePage);

export default ConnectedEditExpensePage;