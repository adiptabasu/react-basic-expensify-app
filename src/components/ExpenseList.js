import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectedExpenses from '../selectors/expenses';

const ExpenseList = (props) => (
    <div>
        {props.expenses.length == 0 ? (
            <p>No expenses</p>
        ) : (
                props.expenses.map((expense) => (
                    <ExpenseListItem
                        key={expense.id}
                        {...expense} />
                ))
            )}
    </div>
);

const mapStateToProps = (state) => {
    return {
        // expenses: state.expenses,
        // filters: state.filters
        expenses: selectedExpenses(state.expenses, state.filters)
    }
}

const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList);

export default ConnectedExpenseList;
export { ExpenseList };