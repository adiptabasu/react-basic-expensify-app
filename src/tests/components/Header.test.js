import { test, expect, jest } from '@jest/globals'
import React from 'react';
// import ReactShallowRenderer from 'react-test-renderer/shallow';
import { shallow } from 'enzyme';
// import toJSON from 'enzyme-to-json'
import { Header } from '../../components/Header';
// test('should render header correctly', () => {
//     const renderer = new ReactShallowRenderer();
//     renderer.render(<Header />);
//     expect(renderer.getRenderOutput()).toMatchSnapshot();
// });

test('should render the header correctly', () => {
    const wrapper = shallow(<Header startLogout={() => { }} />);
    expect(wrapper).toMatchSnapshot();
    //toJSON is not needed, as it is configured to snap in jest.config.json
    // expect(toJSON(wrapper)).toMatchSnapshot();
});

test('should call logout on button click', () => {
    const startLogout = jest.fn();
    const wrapper = shallow(<Header startLogout={startLogout} />);
    wrapper.find('button').simulate('click');
    expect(startLogout).toHaveBeenCalled();
});