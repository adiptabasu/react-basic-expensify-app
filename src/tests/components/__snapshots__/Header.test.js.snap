// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the header correctly 1`] = `
<div>
  <header>
    <h1>
      Expensify
    </h1>
    <NavLink
      className={[Function]}
      to="/"
    >
      Home
    </NavLink>
    <NavLink
      className={[Function]}
      to="/create"
    >
      Create Expense
    </NavLink>
    <NavLink
      className={[Function]}
      to="/edit"
    >
      Edit
    </NavLink>
    <NavLink
      className={[Function]}
      to="/help"
    >
      Help
    </NavLink>
  </header>
</div>
`;
